adapter_ninafw.go

```diff
@@ -212,7 +215,23 @@ func (a *Adapter) startNotifications() {
 	}()
 }
 
-func (a *Adapter) findDevice(handle uint16) Device {
+func (a *Adapter) addDevice(d Device) {
+	a.connectedDevices = append(a.connectedDevices, d)
+}
+
+func (a *Adapter) removeDevice(d Device) {
+	for i := range a.connectedDevices {
+		if d.handle == a.connectedDevices[i].handle {
+			copy(a.connectedDevices[i:], a.connectedDevices[i+1:])
+			a.connectedDevices[len(a.connectedDevices)-1] = Device{} // the zero value of T
+			a.connectedDevices = a.connectedDevices[:len(a.connectedDevices)-1]
```

Comment:
if we don't need to preserve order of the devices, we could just replace the device that needs to be deleted with the last one and reslice the slice to be shorter by one
 ```suggestion
			a.connectedDevices[i] = a.connectedDevices[len(a.connectedDevices)-1]
			a.connectedDevices[len(a.connectedDevices)-1] = Device{} // the zero value of T
			a.connectedDevices = a.connectedDevices[:len(a.connectedDevices)-1]
```
P.S. is setting the last element to zero value in there so that any references in it are released?

Reply:
See https://go.dev/wiki/SliceTricks for why. Trying to avoid possible memory leak.

Reply:
> if we don't need to preserve order of the devices

I think you are correct, we do not need to preserve this order.

Reply:
Updated, thanks for the suggestion!